bindings/wasm/identity_wasm/src/rebased/proposals/borrow.rs

```diff
@@ -0,0 +1,457 @@
+use std::ops::DerefMut;
+use std::rc::Rc;
+use std::result::Result as StdResult;
+
+use identity_iota::iota::rebased::migration::Proposal;
+use identity_iota::iota::rebased::proposals::BorrowAction;
+use identity_iota::iota::rebased::proposals::BorrowIntentFnT;
+use identity_iota::iota::rebased::proposals::ProposalResult;
+use identity_iota::iota::rebased::proposals::ProposalT as _;
+use iota_interaction::rpc_types::IotaObjectData;
+use iota_interaction::types::base_types::ObjectID;
+use iota_interaction::types::programmable_transaction_builder::ProgrammableTransactionBuilder;
+use iota_interaction::types::transaction::Argument;
+use iota_interaction::types::transaction::ProgrammableTransaction;
+use iota_interaction::types::transaction::TransactionKind;
+use iota_interaction_ts::bindings::WasmIotaTransactionBlockEffects;
+use iota_interaction_ts::core_client::WasmCoreClientReadOnly;
+use product_common::bindings::core_client::WasmManagedCoreClientReadOnly;
+use product_common::bindings::transaction::WasmTransactionBuilder;
+use product_common::core_client::CoreClientReadOnly;
+use product_common::transaction::transaction_builder::Transaction as _;
+use product_common::transaction::ProtoTransaction;
+use tokio::sync::RwLock;
+use wasm_bindgen::prelude::*;
+
+use crate::error::Result;
+use crate::error::WasmResult as _;
+use crate::rebased::WasmControllerToken;
+use crate::rebased::WasmOnChainIdentity;
+
+use super::StringSet;
+use super::TransactionDataBuilder;
+
+#[wasm_bindgen(typescript_custom_section)]
+const _TYPE_DEFS: &str = r#"
+import { TransactionDataBuilder, Argument } from "@iota/iota-sdk/transactions";
+import { IotaObjectData } from "@iota/iota-sdk/client";
+
+/**
+ * Closure needed to define what should be done with the borrowed objects.
+ */
+export type BorrowFn = (tx: TransactionDataBuilder, objects: Map<string, [Argument, IotaObjectData]>) => void;
+"#;
+
+#[wasm_bindgen]
+extern "C" {
+  #[derive(Clone)]
+  #[wasm_bindgen(typescript_type = BorrowFn, extends = js_sys::Function)]
+  pub type WasmBorrowFn;
+}
+
+impl WasmBorrowFn {
+  /// The resulting closure may panic if anything goes wrong.
+  /// Make sure to catch_unwind when it's called.
+  pub(crate) fn into_intent_fn(self) -> impl BorrowIntentFnT {
+    type Ptb = ProgrammableTransactionBuilder;
+
+    move |ptb: &mut Ptb, objects: &HashMap<ObjectID, (Argument, IotaObjectData)>| {
+      // Convert the PTB into a TS Transaction (the builder).
+      let pt = std::mem::take(ptb).finish();
+      let tx_kind = TransactionKind::ProgrammableTransaction(pt);
+      let tx_kind_bytes = bcs::to_bytes(&tx_kind).unwrap();
+      let ts_tx_builder = TransactionDataBuilder::from_tx_kind_bcs(tx_kind_bytes).unwrap();
+      // Convert objects into a JS Map of the same types.
+      let ts_map = serde_wasm_bindgen::to_value(objects).unwrap();
+
+      // Call the provided JS closure `borrow_fn`.
+      self
+        .call2(
+          // No `this`.
+          &JsValue::NULL,
```

Comment:
Would this override any `this` in a given `borrow_fn` or is this nested in another closure somewhere?

If yes, should this be documented on the TS function forwarding closures to here?

(Same questions for the other two similar functions. ^^)